### PR TITLE
add option --splinter-webdriver-executable for phantomjs and chrome

### PR DESCRIPTION
**Pull request by sureshvv**

### CODE REVIEW

**Comment by sureshvv** [comment]
> use keyword style for passing such long list of arguments

Are you sure this is a good idea? The arguments already have long names and they are fixtures so cannot make then short, We will end up invoking it like this:

get_args(splinter_webdriver_executavle=splinter_webdriver_executable
splinter_webdriver_executavle=splinter_webdriver_executable
splinter_webdriver_executavle=splinter_webdriver_executable
splinter_webdriver_executavle=splinter_webdriver_executable
splinter_webdriver_executavle=splinter_webdriver_executable
splinter_webdriver_executavle=splinter_webdriver_executable
splinter_webdriver_executavle=splinter_webdriver_executable
splinter_webdriver_executavle=splinter_webdriver_executable
), 

Since this is an internal function anyway, I think it is better to keeo it as it is.

**Comment by bubenkoff** [comment]
nobody stops you to rename the parameters in get_args function so it will
look like:

get_args(executable=splinter_webdriver_executable,
firefox_profile_directory=splinter_firefox_profile_directory...
which is still much better than positional arguments

On 14 April 2015 at 11:01, sureshvv <notifications@github.com> wrote:

> use keyword style for passing such long list of arguments
>
> Are you sure this is a good idea? The arguments already have long names
> and they are fixtures so cannot make then short, We will end up invoking it
> like this:
>
> get_args(splinter_webdriver_executavle=splinter_webdriver_executable
> splinter_webdriver_executavle=splinter_webdriver_executable
> splinter_webdriver_executavle=splinter_webdriver_executable
> splinter_webdriver_executavle=splinter_webdriver_executable
> splinter_webdriver_executavle=splinter_webdriver_executable
> splinter_webdriver_executavle=splinter_webdriver_executable
> splinter_webdriver_executavle=splinter_webdriver_executable
> splinter_webdriver_executavle=splinter_webdriver_executable
> ),
>
> Since this is an internal function anyway, I think it is better to keeo it
> as it is.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pytest-dev/pytest-splinter/pull/24#issuecomment-92705766>
> .
>



-- 
Anatoly Bubenkov

**Comment by sureshvv** [comment]
But then the code inside get_args will have to change, and it will become confusing to people who know/follow the code. Currently git diff now shows the same block of code extracted and neatly moved up into a separate function.
 
While I agree that the keyword style arguments will make the test look prettier, I am not so sure that it is an overall improvement.

**Comment by bubenkoff** [comment]
git diff will also show you renamed arguments, which is pretty readable,
don't worry too much
i worry more about how it looks after refactoring

On 14 April 2015 at 11:35, sureshvv <notifications@github.com> wrote:

> But then the code inside get_args will have to change, and it will become
> confusing to people who know/follow the code. Currently git diff now shows
> the same block of code extracted and neatly moved up into a separate
> function.
>
> While I agree that the keyword style arguments will make the test look
> prettier, I am not so sure that it is an overall improvement.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pytest-dev/pytest-splinter/pull/24#issuecomment-92719485>
> .
>



-- 
Anatoly Bubenkov

File: pytest_splinter/plugin.py [hunk 1]
```diff
@@ -268,22 +307,9 @@ def browser_instance_getter(
...
-                'browser.helperApps.neverAsk.saveToDisk': splinter_download_file_types,
-                'browser.helperApps.alwaysAsk.force': False,
-                'pdfjs.disabled': True,  # disable internal ff pdf viewer to allow auto pdf download
-            }, **splinter_firefox_profile_preferences)
-            kwargs['profile'] = splinter_firefox_profile_directory
-        elif splinter_webdriver == 'remote':
-            kwargs['url'] = splinter_remote_url
-        if splinter_driver_kwargs:
-            kwargs.update(splinter_driver_kwargs)
+        kwargs = get_args(splinter_webdriver, splinter_file_download_dir, splinter_download_file_types,
```
Comment (bubenkoff) [inline pytest_splinter/plugin.py]: did you forget to pass as kwargs?

Comment (sureshvv) [inline pytest_splinter/plugin.py]: Don't follow your question.

get_args returns a dict which is passed as kwargs to the Browser() constructor. get_args has keyword arguments.

Comment (bubenkoff) [inline pytest_splinter/plugin.py]: call get_args like:

```
get_args(driver=splinter_webdriver,....
```

**Comment by bubenkoff** [comment]
thanks!
I'll fix those minor pep257 issues myself
in general it's good idea to set up your editor the way it checks that

**Comment by bubenkoff** [comment]
1.3.5 is out

**Comment by sureshvv** [comment]
Thank you. Is pep257 checking not part of tox tests? I thought I addressed all the issues.

**Comment by bubenkoff** [comment]
now it is :)

On 20 April 2015 at 21:47, sureshvv <notifications@github.com> wrote:

> Thank you. Is pep257 checking not part of tox tests? I thought I addressed
> all the issues.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pytest-dev/pytest-splinter/pull/24#issuecomment-94551672>
> .
>



-- 
Anatoly Bubenkov

**Comment by bh** [comment]
I have an issue after this pull request. My phantomjs binary is in ~/.virtualenvs/mx/bin/ (installed via nodeenv). After update to current version of pytest-splinter, I always get the following exception: 

``` python
    def start(self):
        """
            Starts PhantomJS with GhostDriver.
    
            :Exceptions:
             - WebDriverException : Raised either when it can't start the service
               or when it can't connect to the service
            """
        try:
            self.process = subprocess.Popen(self.service_args, stdin=subprocess.PIPE,
                                            close_fds=platform.system() != 'Windows',
                                            stdout=self._log, stderr=self._log)
    
        except Exception as e:
>           raise WebDriverException("Unable to start phantomjs with ghostdriver.", e)
E           selenium.common.exceptions.WebDriverException: Message: Unable to start phantomjs with ghostdriver.
E           Screenshot: available via screen
```

I fixed this with this workaround in my conftest.py:

``` python
@pytest.fixture(scope='session')
def splinter_webdriver_executable(request):
    return distutils.spawn.find_executable(request.config.option.splinter_webdriver)
```

**Comment by bubenkoff** [comment]
thanks for noticing!
please check out 1.3.6 which is just out

**Comment by sureshvv** [comment]
--splinter-webdriver-executable=$HOME/.virtualenvs/mx/bin should help you. But not sure why this pull request broke it. Was it working before?

**Comment by bubenkoff** [comment]
look at the last commit :)

**Comment by sureshvv** [comment]
aah..  coincidentally named fixtures :)

**Comment by bh** [comment]
On 04/21/2015 01:43 PM, sureshvv wrote:
>
> --splinter-webdriver-executable=$HOME/.virtualenvs/mx/bin should help 
> you. But not sure why this pull request broke it. Was it working before?
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/pytest-dev/pytest-splinter/pull/24#issuecomment-94757873>.
>
Of couse. It should has to be possible to find the binary in $PATH. I 
don't want to be too explicitly ;)

**Comment by bh** [comment]
On 04/21/2015 01:50 PM, Benjamin Hedrich wrote:
> On 04/21/2015 01:43 PM, sureshvv wrote:
>>
>> --splinter-webdriver-executable=$HOME/.virtualenvs/mx/bin should help 
>> you. But not sure why this pull request broke it. Was it working before?
>>
>> —
>> Reply to this email directly or view it on GitHub 
>> <https://github.com/pytest-dev/pytest-splinter/pull/24#issuecomment-94757873>.
>>
> Of couse. It should has to be possible to find the binary in $PATH. I 
> don't want to be too explicitly ;)
py.test --cov=mx --cov-report=term-missing --cov-report html mx/
================================================================================================ 
test session starts 
================================================================================================
platform linux -- Python 3.4.3 -- py-1.4.26 -- pytest-2.7.0
rootdir: /home/bhe/devel/mx/src, inifile: pytest.ini
plugins: splinter, bdd, cov, django, ipdb, rerunfailures, xdist
collected 375 items

mx/account/tests/acceptance/test_email_confirmation.py ..
[...]


With 1.3.6 the issue seems to be resolved! Thank you!

**Comment by sureshvv** [comment]
Actually I think it is better to raise an Exception if executable not found rather than quietly ignore the option,

**Comment by blueyed** [comment]
btw: while looking into why the error from `subprocess.Popen` (the exception) is not
displayed, I've found a bug about the usage of WebDriverException in Selenium's
driver.  See https://github.com/SeleniumHQ/selenium/pull/475.

**Comment by blueyed** [comment]
And I agree with @sureshvv: it's an error and should not pass silently.

**Comment by bubenkoff** [comment]
@sureshvv but where do you see that we silently ignore the option?

**Comment by sureshvv** [comment]
@bubenkoff @blueyed Sorry.. I was reading too fast :-) It is fine as it is

**Pull request merged**